requests/utils.py

```diff
@@ -396,7 +397,28 @@ def unquote_unreserved(uri):
     """Un-escape any percent-escape sequences in a URI that are unreserved
     characters. This leaves all reserved, illegal and non-ASCII bytes encoded.
     """
-    parts = uri.split('%')
+    # This convert function is used to optionally convert the output of `chr`.
+    # In Python 3, `chr` returns a unicode string, while in Python 2 it returns
+    # a bytestring. Here we deal with that by optionally converting.
+    def _convert(is_bytes, c):
+        if is_py2 and not is_bytes:
+            return c.decode('ascii')
+        elif is_py3 and is_bytes:
+            return c.encode('ascii')
+        else:
+            return c
+
+    # Handle both bytestrings and unicode strings.
+    if isinstance(uri, bytes):
+        splitchar = b'%'
+        base = b''
+        convert = functools.partial(_convert, True)
```

Comment:
We could do this another way potentially (although I'm not certain it's necessarily better):

```py
is_bytes = isinstance(uri, bytes)
splitchar = u'%'
base = u''
if is_bytes:
    splitchar = splitchar.encode('ascii')
    base = base.encode('ascii')

# And then down below

if c in UNRESERVED_SET:
    parts[i] = convert(is_bytes, c) + parts[i][2:]
```